README
This is a Flumotion producer component for Blackmagic DeckLink devices.
It is based on the free decklink gstreamer plugin, which is part of
http://gstreamer.freedesktop.org/modules/gst-plugins-bad.html.

Flumotion (http://www.flumotion.net/) is an open source multimedia
streaming server.

BlackMagic (http://www.blackmagicdesign.com/) makes video production hardware.


ChangeLog
=== release 0.2.0 ===

2006-03-06  Thomas Vander Stichele  <thomas at apestaart dot org>

	* NEWS:
	* README:
	* configure.ac:
	  releasing 0.2.0

2006-02-28  Thomas Vander Stichele  <thomas at apestaart dot org>

	* Makefile.am:
	  use an m4/ dir so that autopoint puts its copied macros there
	  instead of cluttering up common

2006-01-27  Thomas Vander Stichele  <thomas at apestaart dot org>

	* flumotion/component/sample/sample.py:
	* flumotion/component/sample/sample.xml:
	  update for new component API

2005-12-20  Thomas Vander Stichele  <thomas at apestaart dot org>

	* flumotion/test/test_sample_admin_gtk.py:
	  after updating setup.m4, we got errors when trial
	  was doing from twisted.internet import reactor
	  in the cleanup phase.  importing it at the top
	  works around it.

2005-12-20  Thomas Vander Stichele  <thomas at apestaart dot org>

	* common/setup.m4:
	  update to register ourselves last
	  update to take a project name
	* configure.ac:
	  add project name to setup call

2005-08-16  Thomas Vander Stichele  <thomas at apestaart dot org>

	* configure.ac:
	  use FLUMOTION_SETUP and our spiffy setup.m4
	* Makefile.am:
	* flumotion/test/Makefile.am:
	  clean FLUMOTION_SETUP files
	* flumotion/test/test_sample_admin_gtk.py:
	* flumotion/test/test_sample_common.py:
	  use setup.setup()

2005-08-08  Thomas Vander Stichele  <thomas at apestaart dot org>

	* configure.ac:
	  update requirement for flumotion to 0.1.9.1 since we use
	  translations and projects setup

2005-07-31  Thomas Vander Stichele  <thomas at apestaart dot org>

	* common/common.mk:
	  enable coverage again
	* flumotion/component/sample/Makefile.am:
	* flumotion/component/sample/common.py:
	  put common methods to both ui and component here
	* flumotion/component/sample/sample.py:
	* flumotion/component/sample/admin_gtk.py:
	* flumotion/component/sample/sample.xml:
	  use common
	* flumotion/test/Makefile.am:
	* flumotion/test/test_sample_admin_gtk.py:
	* flumotion/test/test_sample_common.py:
	  add test for common

2005-07-30  Thomas Vander Stichele  <thomas at apestaart dot org>

	* flumotion/component/sample/admin_gtk.py:
	  reuse more of the base class's code

2005-07-29  Michael Smith <msmith at fluendo.com>

	* flumotion/component/sample/admin_gtk.py:
	  Not finding the locale bundle is a non-fatal error, so that english
	  works.

2005-07-29  Thomas Vander Stichele  <thomas at apestaart dot org>

	* Makefile.am:
	* autogen.sh:
	* common/Makefile.am:
	* common/common.mk:
	* common/gen-locale-xml.py:
	* common/locale.mk:
	* configure.ac:
	* flumotion/component/sample/admin_gtk.py:
	* flumotion/component/sample/sample.glade:
	* flumotion/project/Makefile.am:
	* po/Makevars:
	* po/POTFILES.in:
	* po/fr.po:
	* po/nl.po:
	  add i18n

2005-07-27  Thomas Vander Stichele  <thomas at apestaart dot org>

	* flumotion/component/sample/admin_gtk.py:
	* flumotion/component/sample/sample.glade:
	  add UI and glade
	* flumotion/test/setup.py.in:
	  make good use of os.path.abspath


common/gen-locale-xml.py
# -*- Mode: Python -*-
# vi:si:et:sw=4:sts=4:ts=4

# Flumotion - a streaming media server
# Copyright (C) 2004,2005,2006,2007,2008,2009 Fluendo, S.L.
# Copyright (C) 2010,2011 Flumotion Services, S.A.
# All rights reserved.
#
# This file may be distributed and/or modified under the terms of
# the GNU Lesser General Public License version 2.1 as published by
# the Free Software Foundation.
# This file is distributed without any warranty; without even the implied
# warranty of merchantability or fitness for a particular purpose.
# See "LICENSE.LGPL" in the source distribution for more information.
#
# Headers in this file shall remain intact.

# this script generates a locale registry .xml file for flumotion
# languages are bundled by the two-letter code the locale identifier
# starts with
#
# usage:
# gen-locale-xml.py [gettext domain] [project name] [one or
#                                                    more language codes]

import os
import sys


def main(args):
    if len(sys.argv) < 2:
        sys.stderr.write('Please specify a gettext domain\n')
        sys.exit(1)
    if len(sys.argv) < 3:
        sys.stderr.write('Please specify a project name\n')
        sys.exit(1)
    if len(sys.argv) < 4:
        sys.stderr.write('Please specify one or more language codes\n')
        sys.exit(1)

    domain = sys.argv[1]
    project = sys.argv[2]
    codes = {} # 2-letter language -> full code
    for code in sys.argv[3:]:
        if len(code) < 2:
            sys.stderr.write('Locale code %s is not at least 2 characters\n'
                % code)
            sys.exit(1)
        lang = code[:2]
        if not lang in codes:
            codes[lang] = []
        codes[lang].append(code)

    print "<registry>"
    print
    print "  <bundles>"
    print

    for lang in codes.keys():
        print ("    <bundle name=\"%s-locale-%s\" "
               "under=\"localedatadir\" project=\"%s\">") % (
            domain, lang, project)
        print "      <directories>"
        print '        <directory name="locale">'
        for code in codes[lang]:
            print ("          <filename location"
                   "=\"%s/LC_MESSAGES/%s.mo\" />") % (
                code, domain)
        print "        </directory>"
        print "      </directories>"
        print "    </bundle>"
        print

    print "  </bundles>"
    print
    print "</registry>"

main(sys.argv)


flumotion/worker/checks/decklink.py
# -*- Mode: Python -*-
# vi:si:et:sw=4:sts=4:ts=4

# flumotion-decklink - feeder component for BlackMagic decklink cards
#
# Copyright 2012 Brown University
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA


import gobject
import gst

from flumotion.common import errors, messages

__version__ = "$Rev$"


def getPropsAndCaps():
    """
    Probe the gst decklink plugin for properties and capabilities.

    @returns: L{twisted.internet.defer.Deferred}
    """
    result = messages.Result()

    decklinksrc = gst.element_factory_make('decklinksrc', 'srcprobe')
    if not decklinksrc:
        raise errors.ComponentSetupError('No decklink card found.')

    def get_enum(p):
        vals = []
        if not gobject.type_is_a(p.value_type, gobject.TYPE_ENUM):
            raise errors.ComponentSetupError(p.name + ': unexpected property type.')
        else:
            for item in p.enum_class.__enum_values__.items():
                vals.append({
                    'value': item[0],
                    'name': item[1].value_name,
                    'nick': item[1].value_nick
                })
        return vals

    props = {}
    for p in decklinksrc.props:
        if p.name in ('device', 'subdevice'):
            props['device'] = {
                    'minimum': p.minimum,
                    'maximum': p.maximum
            }
        elif p.name == 'connection':
            props['connection'] = get_enum(p)
        elif p.name == 'mode':
            props['mode'] = get_enum(p)
        elif p.name == 'audio-input':
            props['audio-input'] = get_enum(p)

    def get_data(key, value, props):
        props[key] = value
        return 1

    caps = {
        'video': [],
        'audio': []
    }
    videopad = decklinksrc.get_pad_template('videosrc')
    for cap in videopad.get_caps():
        attrs = { 'name': cap.get_name() }
        cap.foreach(get_data, attrs)
        caps['video'].append(attrs)
    audiopad = decklinksrc.get_pad_template('audiosrc')
    for cap in audiopad.get_caps():
        attrs = { 'name': cap.get_name() }
        cap.foreach(get_data, attrs)
        caps['audio'].append(attrs)

    result.succeed((props, caps))
    return result


flumotion/component/producers/decklink/wizard_gtk.py
# -*- Mode: Python -*-
# vi:si:et:sw=4:sts=4:ts=4

# flumotion-decklink - feeder component for BlackMagic decklink cards
#
# Copyright 2012 Brown University
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA


import gettext
import os

import gobject
import gst

from zope.interface import implements

from flumotion.admin.assistant.interfaces import IProducerPlugin
from flumotion.admin.assistant.models import AudioProducer, VideoProducer,\
	AudioEncoder, VideoEncoder, VideoConverter
from flumotion.common import errors, messages
from flumotion.common.i18n import N_, gettexter
from flumotion.common.messages import Info
from flumotion.admin.gtk.basesteps import AudioProducerStep, VideoProducerStep

__version__ = "$Rev$"
_ = gettext.gettext
T_ = gettexter()


class DecklinkProducer(AudioProducer, VideoProducer):
    componentType = 'decklink-producer'

    def __init__(self):
        super(DecklinkProducer, self).__init__()

        self.properties.device = 0
        self.properties.connection = 0
        self.properties.mode = 0
        self.properties.audio_input = 0
        self.properties.width = 640
        self.properties.height = 360
        self.properties.framerate = 30.0

    def getFeederName(self, component):
        if isinstance(component, AudioEncoder):
            return 'audio'
        elif isinstance(component, (VideoEncoder, VideoConverter)):
            return 'video'
        else:
            raise AssertionError

class _DecklinkCommon:
    # icon = 'decklink.png'
    gladeFile = os.path.join(os.path.dirname(os.path.abspath(__file__)),
                              'wizard.glade')
    componentType = 'decklink'

    def setup(self):
        #self._probeCards()
        self.device.data_type = int
        self.mode.data_type = int
        self.connection.data_type = int
        self.audio_input.data_type = int
        self.width.data_type = int
        self.height.data_type = int
        self.framerate.data_type = float
        self.mode.prefill([('default', self.model.properties.mode)])
        self.connection.prefill([('default', self.model.properties.connection)])
        self.audio_input.prefill([('default', self.model.properties.audio_input)])
        self._aspect = self.model.properties.width / self.model.properties.height;
        self.add_proxy(self.model.properties,
            ['device', 'mode', 'connection', 'audio_input',
             'width', 'height', 'framerate'])
        self._sigs = {
            'mode': self.mode.connect('changed', self.on_mode_changed),
            'width': self.width.connect('value-changed', self.on_width_changed),
            'height': self.height.connect('value-changed', self.on_height_changed)
        }

    def on_mode_changed(self, combo):
        mode = self.mode.get_selected()
        if mode is not None:
            aspect = self.get_aspect(self.model.properties.mode)
            if self._aspect != aspect:
               self._aspect = aspect
               height = int(round(self.model.properties.width / self._aspect))
               if self.model.properties.height != height:
                   self.height.handler_block(self._sigs['height'])
                   self.height.set_value(height)
                   self.height.handler_unblock(self._sigs['height'])

    def on_width_changed(self, entry):
        height = int(round(self.model.properties.width / self._aspect))
        if self.model.properties.height != height:
            self.height.handler_block(self._sigs['height'])
            self.height.set_value(height)
            self.height.handler_unblock(self._sigs['height'])

    def on_height_changed(self, entry):
        width = int(round(self.model.properties.height * self._aspect))
        if self.model.properties.width != width:
            self.width.handler_block(self._sigs['width'])
            self.width.set_value(width)
            self.width.handler_unblock(self._sigs['width'])

    def get_aspect(self, mode):
        cap = self._caps['video'][mode]
        return float(cap['width']) / float(cap['height'])

    def workerChanged(self, worker):
        self.model.worker = worker
        self.wizard.requireElements(worker, 'decklinksrc')
        self._probeCards()

    def _format_menu(self, p):
        menu = []
        for item in p:
            label = item['nick'] or item['name']
            menu.append((str(item['value']) + ": " + label, item['value']))
        return menu

    def _probeCards(self):
        self.vbox1.set_sensitive(False)
        self.wizard.blockNext(True)
        msg = messages.Info(T_(N_('Probing for Decklink devices...')),
            mid='decklink-probe')
        self.wizard.add_msg(msg)

        d = self.wizard.runInWorker(self.model.worker,
            'flumotion.worker.checks.decklink', 'getPropsAndCaps')

        def gotPropsAndCaps((props, caps)):
            self.wizard.clear_msg('decklink-probe')
            self._props = props
            self._caps = caps
            self.mode.prefill(self._format_menu(props['mode']))
            self.connection.prefill(self._format_menu(props['connection']))
            self.audio_input.prefill(self._format_menu(props['audio-input']))
            self._aspect = self.get_aspect(self.model.properties.mode)
            self.wizard.blockNext(False)
            self.vbox1.set_sensitive(True)

        d.addCallback(gotPropsAndCaps)

        return d

class DecklinkVideoStep(_DecklinkCommon, VideoProducerStep):
    name = 'Decklink Video'
    title = _('Decklink Video')
    #docSection = 'help-configuration-assistant-producer-video-decklink'
    #docAnchor = ''

    def __init__(self, wizard, model):
        VideoProducerStep.__init__(self, wizard, model)

    # WizardStep

    def setup(self):
        _DecklinkCommon.setup(self)
        self.audio_frame.hide()

class DecklinkAudioStep(_DecklinkCommon, AudioProducerStep):
    name = 'Decklink Audio'
    title = _('Decklink Audio')
    #docSection = 'help-configuration-assistant-producer-video-decklink'
    #docAnchor = ''

    def __init__(self, wizard, model):
        AudioProducerStep.__init__(self, wizard, model)

    # WizardStep

    def setup(self):
        _DecklinkCommon.setup(self)
        self.video_frame.hide()

    def getNext(self):
        return None

class DecklinkWizardPlugin(object):
    implements(IProducerPlugin)

    def __init__(self, wizard):
        self.wizard = wizard
        self.model = DecklinkProducer()

    def getProductionStep(self, type):
        if type == 'audio':
            return DecklinkAudioStep(self.wizard, self.model)
        elif type == 'video':
            return DecklinkVideoStep(self.wizard, self.model)
